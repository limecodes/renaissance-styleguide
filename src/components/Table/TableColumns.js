import React from 'react'
import styled from 'styled-components'
import { PropTypes } from 'prop-types'

const Column = styled.div`
  border: 1px solid #fff;
  padding: 1rem;
  width: calc(100% / ${(props) => props.columns});
`

const TableColumns = (props) => {
  const { items, columns, numbered, header, number } = props

  if (numbered && header) {
    return (
      <>
        <Column columns={columns + 1}>#</Column>
        {items.map((column, i) => (
          <Column key={`table_column_${i}`} columns={columns + 1}>
            {column || 'N/A'}
          </Column>
        ))}
      </>
    )
  } else if (numbered) {
    return (
      <>
        <Column columns={columns + 1}>{number + 1}</Column>
        {items.map((column, i) => (
          <Column key={`table_column_${i}`} columns={columns + 1}>
            {column || 'N/A'}
          </Column>
        ))}
      </>
    )
  }

  return items.map((column, i) => (
    <Column key={`table_column_${i}`} columns={columns}>
      {column || 'N/A'}
    </Column>
  ))
}

TableColumns.propTypes = {
  columns: PropTypes.number.isRequired,
  items: PropTypes.array.isRequired,
  numbered: PropTypes.bool,
  number: PropTypes.number,
}

export default TableColumns
