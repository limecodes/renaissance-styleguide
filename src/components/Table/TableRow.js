import React from 'react'
import styled from 'styled-components'
import { PropTypes } from 'prop-types'

import TableColumns from './TableColumns'

const Row = styled.div`
  display: flex;
  background-color: #e0dede;
`

const Header = styled(Row)`
  background-color: grey;
  color: #fff;
`

const TableRow = (props) => {
  const { items, columns, header, numbered, number } = props

  return header ? (
    <Header>
      <TableColumns
        header
        numbered={numbered}
        columns={columns}
        items={items}
      />
    </Header>
  ) : (
    <Row>
      <TableColumns
        number={number}
        header={header || false}
        numbered={numbered}
        columns={columns}
        items={items}
      />
    </Row>
  )
}

TableRow.propTypes = {
  items: PropTypes.array.isRequired,
  columns: PropTypes.number.isRequired,
  header: PropTypes.boolean,
  numbered: PropTypes.boolean,
  number: PropTypes.number,
}

export default TableRow
