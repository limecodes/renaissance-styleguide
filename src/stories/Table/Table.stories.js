import React from 'react'

import Table from '../../components/Table'

export default {
  title: 'Common/Table',
  component: Table,
}

const Template = (args) => <Table {...args} />

const headers = [
  'Patient Name',
  'Counselor',
  'Primary Contact',
  'Admission Date',
  'Discharge Date',
]

const patients = [
  {
    name: 'John Marcus Smith',
    counselor: 'James Jacob Gill',
    primaryContact: 'Julia Fay Smith (Spouse)',
    admissionDate: 'January 1st, 1970',
    dischargeDate: null,
  },
  {
    name: 'John Marcus Smith',
    counselor: 'James Jacob Gill',
    primaryContact: 'Julia Fay Smith (Spouse)',
    admissionDate: 'January 1st, 1970',
    dischargeDate: null,
  },
]

export const NoHeaders = Template.bind({})
NoHeaders.args = {
  columns: headers.length,
  contents: patients,
}

export const WithHeaders = Template.bind({})
WithHeaders.args = {
  columns: headers.length,
  contents: patients,
  headers: headers,
}

export const Numbered = Template.bind({})
Numbered.args = {
  columns: headers.length,
  contents: patients,
  headers: headers,
  numbered: true,
}
