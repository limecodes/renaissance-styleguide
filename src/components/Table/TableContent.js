import React from 'react'
import { PropTypes } from 'prop-types'
import styled from 'styled-components'

import TableRow from './TableRow'

const TableContent = (props) => {
  const { items, columns, headers, numbered } = props

  if (headers.length > 0) {
    return (
      <>
        <TableRow
          header
          numbered={numbered}
          columns={columns}
          items={headers}
        />
        {items.map((item, i) => (
          <TableRow
            key={`table_row_${i}`}
            number={numbered ? i : 0}
            numbered={numbered}
            columns={columns}
            items={Object.values(item)}
          />
        ))}
      </>
    )
  }

  return items.map((item, i) => (
    <TableRow
      key={`table_row_${i}`}
      columns={columns}
      items={Object.values(item)}
    />
  ))
}

TableContent.propTypes = {
  columns: PropTypes.number.isRequired,
  items: PropTypes.array.isRequired,
  headers: PropTypes.array,
  numbered: PropTypes.bool,
}

export default TableContent
