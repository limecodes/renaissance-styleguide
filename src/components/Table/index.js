import React from 'react'
import styled from 'styled-components'
import { PropTypes } from 'prop-types'

import TableContent from './TableContent'

const TableWrapper = styled.div`
  display: flex;
  flex-direction: column;
`

/**
- An abstract table that allows headers and displayed in a numbered list
**/

const Table = (props) => {
  const { columns, headers, contents, numbered } = props

  return (
    <TableWrapper>
      <TableContent
        numbered={numbered || false}
        columns={columns}
        items={contents}
        headers={headers || []}
      />
    </TableWrapper>
  )
}

Table.propTypes = {
  /**
    Supply the number of columns to display in the table
   */
  columns: PropTypes.number.isRequired,
  /**
    Provide the headers of the columns for the table, if provided will display the headers of the columns
   */
  headers: PropTypes.array,
  /**
    Provide the contents of the data to be displayed in the table as an array of objects
   */
  contents: PropTypes.array.isRequired,
  /**
    Display the table as a numbered list
   */
  numbered: PropTypes.bool,
}

export default Table
