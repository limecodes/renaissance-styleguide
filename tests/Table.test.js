import React from 'react'
import renderer from 'react-test-renderer'

import Table from '../src/components/Table'

let container

beforeEach(() => {
  container = document.createElement('div')
  document.body.appendChild(container)
})

afterEach(() => {
  document.body.removeChild(container)
  container = null
})

test('Table Renders Successfully', () => {
  const data = ['a', 'b', 'c']

  const tree = renderer.create(<Table columns={3} contents={data} />)

  expect(tree).toMatchSnapshot()
})
