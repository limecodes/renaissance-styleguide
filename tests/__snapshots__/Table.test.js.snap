// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table Renders Successfully 1`] = `
<div
  className="sc-hKFyIo korkGl"
>
  <div
    className="sc-gtssRu kSLPty"
  >
    <div
      className="sc-bdnylx iCXasL"
    >
      a
    </div>
  </div>
  <div
    className="sc-gtssRu kSLPty"
  >
    <div
      className="sc-bdnylx iCXasL"
    >
      b
    </div>
  </div>
  <div
    className="sc-gtssRu kSLPty"
  >
    <div
      className="sc-bdnylx iCXasL"
    >
      c
    </div>
  </div>
</div>
`;
